tensorflow/contrib/rnn/python/ops/rnn_cell.py

```diff
@@ -2095,3 +2096,113 @@ def call(self, inputs, state):
 
     new_state = rnn_cell_impl.LSTMStateTuple(c, m)
     return m, new_state
+
+
+class PrepareableMultiRNNCell(rnn_cell_impl.MultiRNNCell):
+  """Wrapper to implement multi-layer weight normalized RNNs"""
+  def prepare(self):
+    for i, cell in enumerate(self._cells):
+      with vs.variable_scope("cell_%d" % i):
+        cell.prepare()
+
+
+class WeightNormBasicLSTMCell(rnn_cell_impl.BasicLSTMCell):
+  """Weight normalized Basic LSTM Cell.
+
+        The implementation is based on
+        https://arxiv.org/abs/1602.07868
+        Tim Salimans, Diederik P. Kingma
+        Weight Normalization: A Simple Reparameterization to Accelerate
+        Training of Deep Neural Networks
+
+        LSTM implementation is the most basic, non-peephole based.
+  """
+
+  def __init__(self, num_units, norm=True,
+               dtype=dtypes.float32, scope='wn_basic_lstm_cell'):
+    super(WeightNormBasicLSTMCell, self).__init__(num_units,
+                                                  forget_bias=1.0,
+                                                  state_is_tuple=True)
+    self.scope = scope
+    self.dtype = dtype
+    self.prepared = False
+    self.norm = norm
+
+  def prepare(self):
+    """Called by dynamic_rnn just before the time-loop.
+       In this case implements weight-normalization on the weight matrices"""
+    output_size = 4*self._num_units
+
+    # Dynamic RNN requires number of input dimensions = hidden units
+    h_size = x_size = self._num_units
+
+    with vs.variable_scope(self.scope):
+      wx = vs.get_variable("wx", [x_size, output_size], dtype=self.dtype)
+      wh = vs.get_variable("wh", [h_size, output_size], dtype=self.dtype)
+
+      if self.norm:
+        gx = vs.get_variable("gx", [output_size], dtype=self.dtype)
+        gh = vs.get_variable("gh", [output_size], dtype=self.dtype)
+
+        wx = nn_impl.l2_normalize(wx, dim=0) * gx
```

Comment:
Hi @ebrevdo. Thanks for your review. I think this is required in order to make sure the correct tensor names appear on the weights in case the cell is wrapped with lets say a MultiRNNCell with additional variable scopes (cell_%d, for instance). Currently, this happens via the wrapper cell's call() method which is executed for the first time from within dynamic_rnn.

Our first preference was to try and not make any changes to the core tensorflow, but this seemed to be way forward based on the above reasoning. Does this make sense? Or am I missing something here?

Ashwini